caffe2/sgd/adadelta_op.h

```diff
@@ -0,0 +1,179 @@
+#pragma once
+
+#include "caffe2/core/operator.h"
+
+namespace caffe2 {
+
+template <typename Context>
+void adadelta_update(
+    int N,
+    const float* w,
+    const float* g,
+    const float* h,
+    const float* d,
+    float* nw,
+    float* nh,
+    float* nd,
+    float epsilon,
+    float decay,
+    const float* lr,
+    Context* /*context*/) {
+  for (auto i = 0; i < N; ++i) {
```

Comment:
I'd suggest to write this function like.

ConstEigenVectorArrayMap<T> w_arr(w, N);
ConstEigenVectorArrayMap<T> g_arr(g, N);
ConstEigenVectorArrayMap<T> h_arr(h, N);
ConstEigenVectorArrayMap<T> d_arr(d, N);
EigenVectorArrayMap<T> ng_arr(nw, N);
EigenVectorMap<T>(nh, N) = decay * h_arr + (T(1) - decay) * g_arr.square();
ng_arr = (d_arr + epsilon).sqrt() * (h_arr + epsilon).rsqrt() * g_arr;
EigenVectorMap<T>(nd, N) = decay * d_arr + (T(1) - decay) * ng_arr.square();
EigenVectorMap<T>(nw, N) = w_arr + lr[0] * ng_arr;

It will be much faster on CPU to use eigen here since most of the ops above in Eigen have SIMD support.
https://eigen.tuxfamily.org/dox/group__CoeffwiseMathFunctions.html